File: problems/Python/47. Permutations II.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res = []
        path = []
        nums.sort()
        used = [False] * len(nums)

        def bt():
            if len(path) == len(nums):
                res.append(list(path))
                return

            for i in range(len(nums)):
                if used[i] or (i > 0 and nums[i] == nums[i-1] and not used[i-1]):
                    continue
                
                used[i] = True
                path.append(nums[i])
                bt()
                path.pop()
                used[i] = False
        
        bt()
        return res

File: problems/Java/47. Permutations II.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        bt(res, new ArrayList<>(), nums, new boolean[nums.length]);
        return res;
    }

    private void bt(List<List<Integer>> res, List<Integer> path, int[] nums, boolean[] used) {
        if (path.size() == nums.length) {
            res.add(new ArrayList<>(path));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (used[i] || (i > 0 && nums[i] == nums[i-1] && !used[i-1])) {
                continue;
            }
            used[i] = true;
            path.add(nums[i]);
            bt(res, path, nums, used);
            path.remove(path.size() - 1);
            used[i] = false;
        }
    }
}